### PR TITLE
Feature/record timestamps

### PR DESCRIPTION
**Pull request by tam-nguyen-senseye**
* updated condition controlling when onUpdate() is triggered inside Calibration, Plr, and SmoothPursuit

* updated task default names

* updated example app's default tab to "Full Demo"

* updated Video metadata fields and formatting

* added Task model

* refactored TaskData -> EventData

* updated JSON output

[ch17993]

### CODE REVIEW

File: src/api/models/Session.ts [hunk 1]
```diff
@@ -63,23 +63,49 @@ export default class Session {
...
    *
    * @param video A `Video` instance.
    */
-  public async addVideo(video: Models.Video) {
+  public addVideo(video: Models.Video) {
+    if (this.tasks.length === 0) {
+      throw new Error('A Task must be added to the Session before adding any Videos.');
+    }
+
     if (typeof this.metadata.uniqueId === 'string' && this.metadata.uniqueId !== '') {
```
Comment (bobbys-dev) [inline src/api/models/Session.ts:92]: Since you've done this set of checks 3 times already, it might be a candidate function to add in the utils module...? Something like
```
function isNonEmptyString(s) {
    return (typeof s === 'string' && s !== '')
}
```

**Review by bobbys-dev — APPROVED** [review]

File: src/components/tasks/Calibration.tsx [hunk 2]
```diff
@@ -53,8 +53,10 @@ export default function Calibration(props: CalibrationProps) {
   };
 
   React.useEffect(() => {
+    let [prevXIndex, prevYIndex] = [-1, -1];
     const listenerId = moveAnimationValue.addListener((value) => {
-      if (onUpdate) {
+      const { x, y } = value;
+      if (onUpdate && (prevXIndex !== x || prevYIndex !== y) && x === y) {
```
[the remark this answers is not in the document]
Comment (tam-nguyen-senseye) [reply src/components/tasks/Calibration.tsx:59]: yeaa, so that's mainly to prevent redundant calls to `onUpdate()`, which isn't updating the animation, but is a callback meant for recording task data at eventful moments in the animation. Rather than doing that tho, I found that these listeners were triggering on every render during an Animated event (see `Animated.timing()`), e.g. a dot sliding from left to right, or even a dot displaying in the same location for 2s, will produce dozens of onUpdate() calls. The specific condition you pointed out makes it so that `onUpdate()` is only called once per instant the dot is rendered at its updated location. For a more under-the-hood explanation, and if you got time to kill, continue reading below...

To understand how/why this works, you should first take a look at the `interpolate()` function, which takes in an `inputRange` and `outputRange`. Animated uses these values as sort of a reference in order to determine how to convert an input value to its appropriate output value, e.g. for `moveAnimationValue.x.interpolate()`, if inputRange=[0,1] and outputRange=[0,100], then updating moveAnimationValue's x value from 0 to 0.5 will trigger an animation that will move an object horizontally from 0px to 50px . This example defines a linear interpolation, but it can be more or less complex depending on the inputRange/outputRange.

Going back to the Calibration task, we somewhat underutilize this interpolation property, since we only want to display a dot at predetermined points. To do this, we use an inputRange that's simply an array of indices [0, 1, 2, ..., n-1] and map it to an outputRange containing discrete values calculated from `dot_points`. So, when we're updating [x, y], we're actually using the same values for both, e.g. [0, 0] -> [1, 1] -> [2, 2], etc., which then translates to [f(xPos), f(yPos)], where xPos and yPos are obtained from dot_points[0][xPos, yPos] -> dot_points[1][xPos, yPos], etc. Couple this with the `step0` easing function, then the dot will appear to move instantly to its next location with each update to `moveAnimationValue`.

Now, given the fact that [the listeners only provide the updated input values, rather than the interpolated ones](https://animationbook.codedaily.io/add-listener/), and also that Animated doesn't update x and y in unison (perhaps due to their interpolations being calculated separately), you'll begin to see that comparing `x === y` is really just checking to see that both the xPos and yPos are obtained from the same index, and that the dot has finished "moving" to its updated location. 

tl;dr `x` and `y` are a misnomer are being utilized as index values to an array

[the remark this answers is not in the document]
Comment (Haplo-Dragon) [reply src/components/tasks/Calibration.tsx:59]: Oh, that makes a lot more sense! Thanks for taking the time to explain it to me. :smile:

**Review by Haplo-Dragon — APPROVED** [review]
Looks good - nice catch on the util function, @bobbys-dev!
